protocol/block-safety-index.md

```diff
@@ -0,0 +1,224 @@
+- The op-proposer consistency problem
+- The safety indexing problem
+
+At the core, all these issues stem from a single problem:
+determining which L2 block corresponds to which L1 block.
+
+Yet we currently have *different solutions* for each case.
+
+The reason why this is hard: the L2 chain, by design, does not pre-commit to how/where data gets confirmed on L1.
+This reduces reorg risk and DA cost by deferring optionality to the batcher, but introduces "offchain" state.
+
+Can we fix this, and simplify the OP-Stack?
+
+## Sub-problems
+
+### The op-node warmup UX problem
+
+The majority of op-node warmup is derivation-pipeline reset time.
+It traverses to an old L2 block guaranteed to still be canonical,
+then traverses to an old L1 block that is guaranteed to produce
+the inputs needed to continue derivation of next L2 block inputs.
+
+This traversal can be lengthy:
+- The L2 chain is traversed back, because we cannot with certainty say whether it was derived from a recent L1 block.
+  Often this means it "bottoms out" on the finalized L2 block, at least ~12 minutes old.
+- The L1 chain is traversed back a full channel-timeout; 300 blocks, i.e. an hour of L1 data!
+  (reduced in Granite to 50 blocks = 10 minutes).
+
+This delay slows down the process of making nodes fully operational after a restart,
+particularly with regard to derivation.
+
+But it can be so much faster, if we introduce state:
+- L2 traversal can go back to the last L2 block derived from a still-canonical L1 block.
+  Seconds, not minutes, of traversed time-span.
+- L1 traversal can go quickly find the relevant L1 data with a much shorter search space,
+  if we had stricter batch-ordering guarantees (being introduced with Holocene!) and the derived-from mapping.
+    - The last derived-from L1 block is an upper bound.
+    - The previous-to-last derived-from L1 block is a lower bound.
+
+### The span-batch recovery problem
+
+Currently, span-batch recovery (pre-Holocene) relies on 'pending-safe',
+where a span-batch may generate pending blocks that can be reorged out
+if later content of the span-batch is found invalid.
+
+This is changing with Holocene (steady batch derivation, aka strict ordering): we do not want the complexity of having to revert data that was tentatively accepted.
+For interop fault-proving, the L2 assumptions should not be larger than 1 block,
+since we want to consolidate cross-L2 after confirming each L2 optimistically.
+
+If we find a pre-existing L2 chain (simply on node restart, or on reorg),
+we currently cannot be sure it was derived from a certain L1; we have to recreate the derivation state to verify.
+And if the tip was part of a span-batch, we need to find the start of said span-batch.
+So while we can do away with the multi-block pending-safe reorg, we still have to "find" the start of a span-batch.
+If we had an index of L2-block to L1-derived-from, with span-batch bounds info,
+then finding that start would be much faster.
```

Comment:
Don't we need to recreate the derivation state on a reorg or restart anyways? What's the difference to, say, a channel full of singular batches, where the reorg or node restart may have happened in the middle of deriving singular batches from the channel, so the L2 tip is in the middle of a channel? What I mean is that we always have to recreate the derivation state in a way that we get clarity on which L1 block a certain L2 block is derived from, and this will get easier with the set of changes we're introducing with Holocene.

This is not to say that the block safety index is still very useful.